File: tools/selective_lockstep/network_fault_injection.py
from __future__ import annotations

import argparse
import random

from common import load_json, write_result


def main() -> int:
    parser = argparse.ArgumentParser(description="對 encoded-frame manifest 產生 deterministic fault schedule")
    parser.add_argument("manifest")
    parser.add_argument("--seed", type=int, required=True)
    parser.add_argument("--drop-rate", type=float, default=0.0)
    parser.add_argument("--duplicate-rate", type=float, default=0.0)
    parser.add_argument("--output")
    args = parser.parse_args()
    if not 0 <= args.drop_rate <= 1 or not 0 <= args.duplicate_rate <= 1:
        parser.error("rates must be in [0, 1]")
    frames = load_json(args.manifest)["frames"]
    rng = random.Random(args.seed)
    schedule = []
    for frame in frames:
        roll = rng.random()
        action = "drop" if roll < args.drop_rate else "duplicate" if roll < args.drop_rate + args.duplicate_rate else "deliver"
        schedule.append({"team_sequence": frame["team_sequence"], "action": action})
    write_result(args.output, {"seed": args.seed, "schedule": schedule})
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/selective_lockstep/network_fault_injection.lua
local dir=debug.getinfo(1,'S').source:sub(2):match('^(.*)[/\\]');package.path=dir..'/?.lua;'..package.path
local common=require('common');local o=common.options(arg);assert(o.positional[1],'manifest required');local seed=assert(math.tointeger(tonumber(o.seed)),'--seed required');local drop=tonumber(o.drop_rate or 0);local duplicate=tonumber(o.duplicate_rate or 0);assert(drop>=0 and drop<=1 and duplicate>=0 and duplicate<=1,'rates must be in [0, 1]')
-- Stable local PRNG. Its sequence is part of the Lua tool contract.
local state=seed&0xffffffff;local function random()state=(1664525*state+1013904223)&0xffffffff;return state/0x100000000 end
local schedule={};for _,frame in ipairs(common.load_json(o.positional[1]).frames)do local roll=random();local action=roll<drop and'drop'or roll<drop+duplicate and'duplicate'or'deliver';table.insert(schedule,{team_sequence=frame.team_sequence,action=action})end
common.write_result(o.output,{seed=seed,schedule=schedule})
